UefiTestingPkg/Library/FlatPageTableLib/AArch64/FlatPageTableAArch64.c

```diff
@@ -0,0 +1,371 @@
+/** @file
+  AArch64 specific page table attribute library functions.
+
+  Copyright (c) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Uefi.h>
+#include <Library/ArmLib.h>
+#include <Chipset/AArch64Mmu.h>
+#include <Library/DebugLib.h>
+#include <Library/FlatPageTableLib.h>
+
+#define TCR_EL1_HPD_FIELD             BIT41   // Assumes translation table is located at TTBR0 (UEFI spec dictated)
+#define TCR_EL2_HPD_FIELD             BIT24
+#define ID_AA64MMFR1_EL1_HPD_MASK     0xF000
+#define TT_HERITABLE_ATTRIBUTES_MASK  (TT_TABLE_AP_MASK | TT_TABLE_PXN | TT_TABLE_UXN)
+
+#define IS_VALID(page)                 ((page & 0x1) != 0)
+#define IS_TABLE(page, level)          ((level == 3) ? FALSE : (((page) & TT_TYPE_MASK) == TT_TYPE_TABLE_ENTRY))
+#define IS_BLOCK(page, level)          ((level == 3) ? (((page) & TT_TYPE_MASK) == TT_TYPE_BLOCK_ENTRY_LEVEL3) : ((page & TT_TYPE_MASK) == TT_TYPE_BLOCK_ENTRY))
+#define ROOT_TABLE_LEN(T0SZ)           (TT_ENTRY_COUNT >> ((T0SZ) - 16) % 9)
+#define ARM_TT_BASE_ADDRESS(page)      (page & TT_ADDRESS_MASK_BLOCK_ENTRY)
+#define ARM_TT_BLOCK_ATTRIBUTES(page)  (page & AARCH64_ATTRIBUTES_MASK)
+
+typedef union {
+  struct {
+    UINT64    Valid           : 1;  // BIT0
+    UINT64    BlockOrTable    : 1;  // BIT1
+    UINT64    LowerAttributes : 10; // BIT2-11
+    UINT64    Address         : 36; // BIT12-47
+    UINT64    Reserved        : 3;  // BIT48-50
+    UINT64    Ignored         : 8;  // BIT51-58
+    UINT64    PxnTable        : 1;  // BIT59
+    UINT64    XnTable         : 1;  // BIT60
+    UINT64    ApTable         : 2;  // BIT61-62
+    UINT64    NsTable         : 1;  // BIT63
+  } Bits;
+  UINT64    Uint64;
+} TRANSLATION_TABLE_ENTRY_HERITABLE;
+
+typedef union {
+  struct {
+    UINT64    Valid           : 1;  // BIT0
+    UINT64    BlockOrTable    : 1;  // BIT1
+    UINT64    LowerAttributes : 10; // BIT2-11
+    UINT64    LevelZeroIndex  : 9;  // BIT12-20
+    UINT64    LevelOneIndex   : 9;  // BIT21-29
+    UINT64    LevelTwoIndex   : 9;  // BIT30-38
+    UINT64    LevelThreeIndex : 9;  // BIT39-47
+    UINT64    Reserved        : 3;  // BIT48-50
+    UINT64    Ignored         : 8;  // BIT51-58
+    UINT64    PxnTable        : 1;  // BIT59
+    UINT64    XnTable         : 1;  // BIT60
+    UINT64    ApTable         : 2;  // BIT61-62
+    UINT64    NsTable         : 1;  // BIT63
+  } Bits;
+  UINT64    Uint64;
+} TRANSLATION_TABLE_ENTRY_TABLE;
+
+typedef AARCH64_PAGE_MAP_ENTRY TRANSLATION_TABLE_ENTRY_BLOCK;
+
+typedef union {
+  TRANSLATION_TABLE_ENTRY_BLOCK        Tteb;
+  TRANSLATION_TABLE_ENTRY_TABLE        Ttet;
+  TRANSLATION_TABLE_ENTRY_HERITABLE    Tteh;
+  UINT64                               Uint64;
+} TRANSLATION_TABLE_ENTRY_UNION;

```

Comment:
defining the source of these structures usually helps in the future and/or when arch makes updates known the impact of such an update.  are thse defined in arm spec?  what revision?

Reply:
added